### ISSUE TITLE
About Cluster Process

### ISSUE THREAD

**Issue by ycwfs** [mjkwon2021/SAFIRE#17]
Dear Author:
I wonder if the following statements are correct:
       1、cluster is a Non-differentiable step, so it is not applied during training
       2、cluster is only used in the multi-inference process (By cluster mid-feature to distinguish different regions, the regions contain forged and authentic regions)

So if I only want multi-regions that have been forged, should i combine the predicted probability and multi-inference result to obtain.

**Comment by CauchyComplete** [comment]
Hi,

1. Clustering is used only at inference time. Not in pretraining or training.
2. Clustering is used for both binary and multi cases. For binary, k is fixed to 2.

> So if I only want multi-regions that have been forged, should i combine the predicted probability and multi-inference result to obtain.

Well, what is the predicted probability here? SAFIRE does not provide that value. I don't understand this sentence.

**Comment by ycwfs** [comment]
anns, safire_pred, max_confidence_indices = safire_automatic_model.safire_predict(npimage)

                ann = {
                    "segmentation": mask_data["segmentations"][idx],
                    "area": area_from_rle(mask_data["rles"][idx]),
                    "bbox": box_xyxy_to_xywh(mask_data["boxes"][idx]).tolist(),
                    "predicted_iou": mask_data["iou_preds"][idx].item(),
                    "point_coords": [mask_data["points"][idx].tolist()],
                    "stability_score": mask_data["stability_score"][idx].item(),
                    "pred_mask": mask_data["pred_masks"][idx],
                }

safire_pred is the predicted probability, meaning the likelihood of each pixel being forged?? Or any value in the anns??

**Comment by ycwfs** [comment]
So the model only outputs the predicted mask matrix, without confidence scores?

**Comment by ycwfs** [comment]
<img width="1958" height="791" alt="Image" src="https://github.com/user-attachments/assets/bb6de8d8-d7ec-4687-b3b8-d3d3ffed3795" />

For example, the purple region is the forged region, the orange and green regions are authentic, but in multi-inference, the cluster can't distinguish whether the region is forged.

**Comment by CauchyComplete** [comment]
Please read the paper. There is no distinction among sources.
